source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -84,32 +87,43 @@ class GrpcMuxFailover : public GrpcStreamInterface<RequestType, ResponseType>,
   // Attempts to establish a new stream to the either the primary or failover source.
   void establishNewStream() override {
     // Attempt establishing a connection to the primary source.
-    // This method may be called multiple times, even if the primary stream is already
-    // established or in the process of being established.
+    // This method may be called multiple times, even if the primary/failover stream
+    // is already established or in the process of being established.
     if (complete_retry_timer_) {
       complete_retry_timer_->disableTimer();
     }
+    // If already connected to one of the source, return.
+    if (connection_state_ == ConnectionState::ConnectedToPrimary ||
+        connection_state_ == ConnectionState::ConnectedToFailover) {
+      ENVOY_LOG_MISC(trace,
+                     "Already connected to an xDS server, skipping establishNewStream() call");
+      return;
+    }
     // First check if Envoy ever connected to the primary/failover, and if so
     // persist attempts to that source.
-    if (ever_connected_to_ == ConnectedState::Primary) {
-      ASSERT(!connecting_to_failover_);
+    // Prior connection to failover is disregarded as Envoy will try
+    // reconnecting to the primary again.
+    if (ever_connected_to_primary_) {
```

Comment:
why do we need `ever_connected_to_primary_` field? Would it not be enough just to try to connect to primary by default if `connection_state_ == ConnectionState::None || connection_state_ == ConnectionState::ConnectingToPrimary`?

Reply:
Good question.
The requirement is to be never connect to the failover source after Envoy was connected to the primary.
Note that the `ever_connected_to_primary_` is also used in [line 219](https://github.com/envoyproxy/envoy/blob/1f53139b64319a903170c3ba190ccc61bc4dec42/source/extensions/config_subscription/grpc/grpc_mux_failover.h#L219). There it is used as a marker to remember not to reconnect to the primary, so the flag is still needed.

In the context of this method, however, I agree that it is redundant, and having less code is probably better in this case. I've removed this piece of code, and added a comment explaining why the requirement holds. Thanks for pointing this out!